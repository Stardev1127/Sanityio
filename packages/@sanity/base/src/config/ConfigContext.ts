import {createContext} from 'react'
import {SanityConfig} from './types'

export const ConfigContext = createContext<SanityConfig | null>(null)
