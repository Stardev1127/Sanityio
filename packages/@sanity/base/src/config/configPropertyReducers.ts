import {Template, TemplateResponse} from '../templates'
import {
  DocumentActionComponent,
  DeleteAction,
  DiscardChangesAction,
  DuplicateAction,
  PublishAction,
  UnpublishAction,
  DocumentBadgeComponent,
  LiveEditBadge,
} from '../deskTool'
import {
  ConfigContext,
  ConfigPropertyReducer,
  AsyncConfigPropertyReducer,
  ResolveProductionUrlContext,
  SanityTool,
  DocumentActionsContext,
  DocumentBadgesContext,
  NewDocumentOptionsContext,
} from './types'

export const initialDocumentBadges = [LiveEditBadge]

export const initialDocumentActions = [
  PublishAction,
  DiscardChangesAction,
  UnpublishAction,
  DuplicateAction,
  DeleteAction,
]

export const schemaTypesReducer: ConfigPropertyReducer<
  unknown[],
  Omit<ConfigContext, 'schema' | 'currentUser'>
> = (prev, {schema}, context) => {
  const schemaTypes = schema?.types

  if (!schemaTypes) return prev
  if (typeof schemaTypes === 'function') return schemaTypes(prev, context)
  if (Array.isArray(schemaTypes)) return [...prev, ...schemaTypes]

  throw new Error(
    `Expected \`schema.types\` to an array or a function but found ${typeof schemaTypes} instead.`
  )
}

export const resolveProductionUrlReducer: AsyncConfigPropertyReducer<
  string | undefined,
  ResolveProductionUrlContext
> = async (prev, {document}, context) => {
  const resolveProductionUrl = document?.resolveProductionUrl
  // the redundant await is useful for error logging because the error is caught
  // in this stack vs somewhere down stream
  // eslint-disable-next-line no-return-await
  if (resolveProductionUrl) return await resolveProductionUrl(prev, context)
  return prev
}

export const toolsReducer: ConfigPropertyReducer<SanityTool[], ConfigContext> = (
  prev,
  {tools},
  context
) => {
  if (!tools) return prev
  if (typeof tools === 'function') return tools(prev, context)
  if (Array.isArray(tools)) return [...prev, ...tools]

  throw new Error(`Expected \`tools\` to an array or a function but found ${typeof tools} instead.`)
}

export const schemaTemplatesReducer: ConfigPropertyReducer<Template[], ConfigContext> = (
  prev,
  {schema},
  context
) => {
  const schemaTemplates = schema?.templates
  if (!schemaTemplates) return prev
  if (typeof schemaTemplates === 'function') return schemaTemplates(prev, context)
  if (Array.isArray(schemaTemplates)) return [...prev, ...schemaTemplates]

  throw new Error(
    `Expected \`schema.templates\` to an array or a function but found ${typeof schemaTemplates} instead.`
  )
}

export const documentBadgesReducer: ConfigPropertyReducer<
  DocumentBadgeComponent[],
  DocumentBadgesContext
> = (prev, {document}, context) => {
  const documentBadges = document?.badges
  if (!documentBadges) return prev

  if (typeof documentBadges === 'function') return documentBadges(prev, context)
  if (Array.isArray(documentBadges)) return [...prev, ...documentBadges]

  throw new Error(
    `Expected \`document.actions\` to an array or a function but found ${typeof documentBadges} instead.`
  )
}

export const documentActionsReducer: ConfigPropertyReducer<
  DocumentActionComponent[],
  DocumentActionsContext
> = (prev, {document}, context) => {
  const documentActions = document?.actions
  if (!documentActions) return prev

  if (typeof documentActions === 'function') return documentActions(prev, context)
  if (Array.isArray(documentActions)) return [...prev, ...documentActions]

  throw new Error(
    `Expected \`document.actions\` to an array or a function but found ${typeof documentActions} instead.`
  )
}

export const newDocumentOptionsResolver: ConfigPropertyReducer<
  TemplateResponse[],
  NewDocumentOptionsContext
> = (prev, {document}, context) => {
  const resolveNewDocumentOptions = document?.resolveNewDocumentOptions
  if (!resolveNewDocumentOptions) return prev

  if (typeof resolveNewDocumentOptions !== 'function') {
    throw new Error(
      `Expected \`document.resolveNewDocumentOptions\` a function but found ${typeof resolveNewDocumentOptions} instead.`
    )
  }

  return resolveNewDocumentOptions(prev, context)
}
