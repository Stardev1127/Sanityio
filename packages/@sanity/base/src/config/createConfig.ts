import {WorkspaceOptions, Config} from './types'

export function createConfig(config: WorkspaceOptions | WorkspaceOptions[]): Config {
  return {type: 'sanity-config', __internal: config}
}
