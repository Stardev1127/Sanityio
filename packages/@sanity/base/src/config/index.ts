export * from './createConfig'
export * from './createPlugin'
export * from './types'
export * from './resolveConfig'
export * from './ConfigPropertyError'
export * from './ConfigResolutionError'
export * from './SchemaError'
