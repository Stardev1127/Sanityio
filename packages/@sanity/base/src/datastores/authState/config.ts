/* eslint-disable camelcase */

export interface AuthStateConfig {
  readConfig: () => {projectId: string}
  authTokenIsAllowed: () => boolean
}

export function __tmp_authState_config(context: {
  loginConfig?: {loginMethod: 'dual' | 'cookie'}
  projectId: string
}) {
  const {loginConfig, projectId} = context

  const loginMethod = ['dual', 'cookie'].includes(loginConfig?.loginMethod || '')
    ? loginConfig?.loginMethod
    : 'dual'

  // export
  function readConfig(): {projectId: string} {
    return {projectId}
  }

  // export
  function authTokenIsAllowed(): boolean {
    return loginMethod === 'dual'
  }

  return {readConfig, authTokenIsAllowed}
}
