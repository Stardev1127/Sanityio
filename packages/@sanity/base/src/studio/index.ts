export * from './StudioProvider'
export * from './Studio'
export * from './renderStudio'
export * from './source'
export * from './workspace'
export * from './config'
export * from './location'
export * from './colorScheme'
