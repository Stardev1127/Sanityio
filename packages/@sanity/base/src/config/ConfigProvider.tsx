import React from 'react'
import {SanityConfig} from './types'
import {ConfigContext} from './ConfigContext'

export function ConfigProvider(props: {children?: React.ReactNode; config: SanityConfig}) {
  const {children, config} = props

  return <ConfigContext.Provider value={config}>{children}</ConfigContext.Provider>
}
