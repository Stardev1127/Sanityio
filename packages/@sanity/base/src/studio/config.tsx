import React, {createContext, useContext} from 'react'
import {Config} from '../config'

interface ConfigProviderProps {
  config: Config
  children?: React.ReactChild
}
const ConfigContext = createContext<Config | null>(null)
export function ConfigProvider({children, config}: ConfigProviderProps) {
  return <ConfigContext.Provider value={config}>{children}</ConfigContext.Provider>
}
export function useConfig() {
  const config = useContext(ConfigContext)
  if (!config) throw new Error('Could not find `config` context')
  return config
}
