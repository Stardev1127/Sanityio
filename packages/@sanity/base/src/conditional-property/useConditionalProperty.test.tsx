import {renderHook} from '@testing-library/react-hooks'
import React, {useMemo} from 'react'
// import {createMockSanityClient} from '../../test/mocks/mockSanityClient'
import {createConfig} from '../config'
import {StudioProvider} from '../studio'
import {
  ConditionalPropertyProps,
  unstable_useConditionalProperty as useConditionalProperty,
} from './useConditionalProperty'

// This mock is needed to prevent the "not wrapped in act()" error from React testing library.
// The reason is that the `useCurrentUser` is used by `ObjectInput` to figure out which fields are
// hidden, and using this hook causes the `ObjectInput` to render again once the user is loaded.
//
// NOTE!
// We can remove this mock when `ObjectInput` no longer uses `useCurrentUser`.
jest.mock('../datastores/user/hooks', () => {
  const hooks = jest.requireActual('../datastores/user/hooks')

  return {
    ...hooks,
    useCurrentUser: jest.fn().mockImplementation(() => ({
      value: {},
      error: null,
      isLoading: false,
    })),
  }
})

const dummyDocument = {
  _createdAt: '2021-11-04T15:41:48Z',
  _id: 'drafts.10053a07-8647-4ebd-9d1d-33a512d30d3a',
  _rev: '5hb8s6-k75-ip4-4bq-5ztbf3fbx',
  _type: 'conditionalFieldsTest',
  _updatedAt: '2021-11-05T12:34:29Z',
  title: 'Hello world',
  isPublished: true,
}

const DEFAULT_PROPS: Omit<ConditionalPropertyProps, 'checkProperty'> = {
  checkPropertyKey: 'testKey',
  document: dummyDocument,
  value: undefined,
  parent: {
    parentTest: 'hello',
  },
}

function TestWrapper({children}: any) {
  const config = useMemo(
    () =>
      createConfig({
        // clientFactory: () => createMockSanityClient() as any,
        name: 'test',
        title: 'Test',
        projectId: 'foo',
        dataset: 'test',
      }),
    []
  )

  return <StudioProvider config={config}>{children}</StudioProvider>
}

afterEach(() => {
  jest.resetAllMocks()
})

describe('Conditional property resolver', () => {
  it('calls callback function', () => {
    const callbackFn = jest.fn(() => true)

    const {result} = renderHook(
      () =>
        useConditionalProperty({
          checkProperty: callbackFn,
          ...DEFAULT_PROPS,
        }),
      {wrapper: TestWrapper}
    )

    if (result.error) {
      throw result.error
    }

    expect(callbackFn).toBeCalled()

    expect(callbackFn.mock.calls).toMatchSnapshot()
  })

  it('resolves callback to true', () => {
    const {result} = renderHook(
      () =>
        useConditionalProperty({
          // eslint-disable-next-line max-nested-callbacks
          checkProperty: jest.fn(() => true),
          ...DEFAULT_PROPS,
        }),
      {wrapper: TestWrapper}
    )
    expect(result.current).toBeTruthy()
  })

  it('returns false with callback that returns false', () => {
    const {result} = renderHook(
      () =>
        useConditionalProperty({
          // eslint-disable-next-line max-nested-callbacks
          checkProperty: jest.fn(() => false),
          ...DEFAULT_PROPS,
        }),
      {wrapper: TestWrapper}
    )
    expect(result.current).toBe(false)
  })

  it('returns false if document title does not match', () => {
    const {result} = renderHook(
      () =>
        useConditionalProperty({
          // eslint-disable-next-line max-nested-callbacks
          checkProperty: jest.fn(({document}) => document?.title !== 'Hello world'),
          ...DEFAULT_PROPS,
        }),
      {wrapper: TestWrapper}
    )
    expect(result.current).toBeFalsy()
  })

  it('returns true if document is published', () => {
    const {result} = renderHook(
      () =>
        useConditionalProperty({
          // eslint-disable-next-line max-nested-callbacks
          checkProperty: jest.fn(({document}) => Boolean(document?.isPublished)),
          ...DEFAULT_PROPS,
        }),
      {wrapper: TestWrapper}
    )
    expect(result.current).toBeTruthy()
  })

  it('returns undefined because callback returns undefined', () => {
    const {result} = renderHook(
      () =>
        useConditionalProperty({
          // eslint-disable-next-line max-nested-callbacks
          checkProperty: jest.fn(() => undefined) as any,
          ...DEFAULT_PROPS,
        }),
      {wrapper: TestWrapper}
    )
    expect(result.current).toBe(undefined)
  })

  it('returns true because value matches', () => {
    const {result} = renderHook(
      () =>
        useConditionalProperty({
          // eslint-disable-next-line max-nested-callbacks
          checkProperty: jest.fn(({value}) => value === 'test value'),
          ...DEFAULT_PROPS,
          value: 'test value',
        }),
      {wrapper: TestWrapper}
    )
    expect(result.current).toBeTruthy()
  })

  it('returns false because value does not match', () => {
    const {result} = renderHook(
      () =>
        useConditionalProperty({
          // eslint-disable-next-line max-nested-callbacks
          checkProperty: jest.fn(({value}) => value === 'test'),
          ...DEFAULT_PROPS,
          value: 'test value',
        }),
      {wrapper: TestWrapper}
    )
    expect(result.current).toBeFalsy()
  })

  it.todo('returns true when current user role is not administrator')
  it.todo('returns true when sibling field is not empty')
})
