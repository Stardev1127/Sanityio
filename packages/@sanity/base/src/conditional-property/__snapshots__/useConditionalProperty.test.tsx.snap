// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Conditional property resolver calls callback function 1`] = `
Array [
  Array [
    Object {
      "currentUser": Object {},
      "document": Object {
        "_createdAt": "2021-11-04T15:41:48Z",
        "_id": "drafts.10053a07-8647-4ebd-9d1d-33a512d30d3a",
        "_rev": "5hb8s6-k75-ip4-4bq-5ztbf3fbx",
        "_type": "conditionalFieldsTest",
        "_updatedAt": "2021-11-05T12:34:29Z",
        "isPublished": true,
        "title": "Hello world",
      },
      "parent": Object {
        "parentTest": "hello",
      },
      "value": undefined,
    },
  ],
]
`;
